nep-17.mediawiki

```diff
@@ -0,0 +1,107 @@
+
+==Specification==
+
+In the method definitions below, we provide both the definitions of the functions as they are defined in the contract as well as the invoke parameters.
+
+===Methods===
+
+====symbol====
+
+<pre>
+public static string symbol()
+</pre>
+
+Returns a short string symbol of the token managed in this contract. e.g. <code>"MYT"</code>. This symbol SHOULD be short (3-8 characters is recommended), with no whitespace characters or new-lines and SHOULD be limited to the uppercase latin alphabet (i.e. the 26 letters used in English).
+
+This method MUST always return the same value every time it is invoked.
+
+====decimals====
+
+<pre>
+public static byte decimals()
+</pre>
+
+Returns the number of decimals used by the token - e.g. <code>8</code>, means to divide the token amount by <code>100,000,000</code> to get its user representation.
+
+This method MUST always return the same value every time it is invoked.
+
+====totalSupply====
+
+<pre>
+public static BigInteger totalSupply()
+</pre>
+
+Returns the total token supply deployed in the system.
+
+====balanceOf====
+
+<pre>
+public static BigInteger balanceOf(byte[] account)
+</pre>
+
+Returns the token balance of the <code>account</code>.
+
+The parameter <code>account</code> MUST be a 20-byte address. If not, this method SHOULD <code>throw</code> an exception.
+
+If the <code>account</code> is an unused address, this method MUST return <code>0</code>.
+
+====transfer====
+
+<pre>
+public static bool transfer(byte[] from, byte[] to, BigInteger amount)
+</pre>
+
+Transfers an <code>amount</code> of tokens from the <code>from</code> account to the <code>to</code> account.
+
+The parameters <code>from</code> and <code>to</code> MUST be 20-byte addresses. If not, this method SHOULD <code>throw</code> an exception.
+
+The parameter <code>amount</code> MUST be greater than or equal to <code>0</code>. If not, this method SHOULD <code>throw</code> an exception.
+
+The function MUST return <code>false</code> if the <code>from</code> account balance does not have enough tokens to spend.
+
+If the method succeeds, it MUST fire the <code>transfer</code> event, and MUST return <code>true</code>, even if the <code>amount</code> is <code>0</code>, or <code>from</code> and <code>to</code> are the same address.
+
+The function SHOULD check whether the <code>from</code> address equals the caller contract hash. If so, the transfer SHOULD be processed; If not, the function SHOULD use the SYSCALL <code>Neo.Runtime.CheckWitness</code> to verify the transfer.
+
+If the transfer is not processed, the function MUST return <code>false</code>.
+
+The function MUST call <code>onPayment</code> method AFTER fire the <code>Transfer</code> event if the receiver is a deployed contract. If the receiver doesn't want to receive this transfer it MUST call <code>ABORT</code>.
```

Comment:
```suggestion
If the receiver is a deployed contract, the function MUST call <code>onPayment</code> method on receiver contract AFTER firing the <code>Transfer</code> event. If the receiver doesn't want to receive this transfer it MUST call <code>ABORT</code>.
```

Reply:
I would also add: "If onPayment method does not exist on receiver contract, no action is required."

Reply:
I'd rather fail if `onPayment` does not exist. `onPayment` is mandatory for contracts that want to receive/own/spend tokens IMO.

Reply:
This would make deployed need to update its rules over time or make always true. 
I am not sure about make mandatory.

Reply:
I am thinking again here, the default will be Abort anyway,right?
Thus if not implemented it will not accept tokens anyway.

Reply:
> I'd rather fail if onPayment does not exist. onPayment is mandatory for contracts that want to receive/own/spend tokens IMO.

Looks like a circular/reflexive argument to me... 
- we fail if method doesn't exist, because onPayment is mandatory to receive tokens.. and since it's mandatory, we should fail if it doesn't exist :) 
- We could not fail if it doesn't exist, and not make mandatory the existance of onPayment to receive tokens.. and since it's not mandatory, we are obliged to fail if it doesn't exist (opposite logic)

As long as this is clear, we should write on text anyway, the preferred logic.

I think this is a very important matter, regarding future standards, that will certainly mimic this. A deployed contract released today should be unable to receive some token kind developed in 5 years from now? Example: NFT will have its own "onPayment" strategy, so if it's not predicted on contract, it will not be able to receive it.

Some more general "strategy" could be a method "allowReceivingAssets" or typical "payable" flag. Maybe, the sense of "payable" is still valid, and we should require contracts to be "payable" (in manifest) and call NEP-5 "onPayment" following its own structure (for NFT it's different, and so on). Just to make sure that if we follow this logic here, we will block any future developed token kinds to be received by any contract developed in this moment (something that a complementary strategy of "payable=true" would resolve for any future "token kinds").

Reply:
A more flexible structure to me for "payable assets" on general, seems to be:
- contract wants to send assets (NEP-5, NFT, ...) to another contract
- if contract is not deployed, it will accept
- if contract is deployed, check if it's payable
- if it's not payable, reject.
- if it's payable, send assets and call specific action method (for NEP-5 is "onPayment")